service/history/transferQueueActiveTaskExecutor.go

```diff
@@ -207,73 +206,54 @@ func (t *transferQueueActiveTaskExecutor) processActivityTask(
 
 func (t *transferQueueActiveTaskExecutor) processWorkflowTask(
 	ctx context.Context,
-	task *tasks.WorkflowTask,
+	transferTask *tasks.WorkflowTask,
 ) (retError error) {
 	ctx, cancel := context.WithTimeout(ctx, taskTimeout)
 	defer cancel()
 
-	weContext, release, err := getWorkflowExecutionContextForTask(ctx, t.cache, task)
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, t.cache, transferTask)
 	if err != nil {
 		return err
 	}
 	defer func() { release(retError) }()
 
-	mutableState, err := loadMutableStateForTransferTask(ctx, weContext, task, t.metricHandler, t.logger)
+	mutableState, err := loadMutableStateForTransferTask(ctx, weContext, transferTask, t.metricHandler, t.logger)
 	if err != nil {
 		return err
 	}
 	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
 		return nil
 	}
 
-	workflowTask := mutableState.GetWorkflowTaskByID(task.ScheduledEventID)
+	workflowTask := mutableState.GetWorkflowTaskByID(transferTask.ScheduledEventID)
 	if workflowTask == nil {
 		return nil
 	}
-	err = CheckTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), workflowTask.Version, task.Version, task)
+	err = CheckTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), workflowTask.Version, transferTask.Version, transferTask)
 	if err != nil {
 		return err
 	}
 
-	executionInfo := mutableState.GetExecutionInfo()
+	// Task queue from transfer task (not current one from mutable state) must be used here.
+	// If current task queue becomes sticky since this transfer task was created,
+	// it can't be used here, because timeout timer was not created for it,
+	// because it used to be non-sticky when this transfer task was created .
+	taskQueue, scheduleToStartTimeout := mutableState.TaskQueueScheduleToStartTimeout(transferTask.TaskQueue)
 
-	// NOTE: previously this section check whether mutable state has enabled
-	// sticky workflowTask, if so convert the workflowTask to a sticky workflowTask.
-	// that logic has a bug which timer task for that sticky workflowTask is not generated
-	// the correct logic should check whether the workflow task is a sticky workflowTask
-	// task or not.
-	var taskQueue *taskqueuepb.TaskQueue
-	var taskScheduleToStartTimeout *time.Duration
-	if mutableState.GetExecutionInfo().TaskQueue != task.TaskQueue {
-		// this workflowTask is an sticky workflowTask
-		// there shall already be an timer set
-		taskQueue = &taskqueuepb.TaskQueue{
-			Name: task.TaskQueue,
-			Kind: enumspb.TASK_QUEUE_KIND_STICKY,
-		}
-		taskScheduleToStartTimeout = executionInfo.StickyScheduleToStartTimeout
-	} else {
-		taskQueue = &taskqueuepb.TaskQueue{
-			Name: task.TaskQueue,
-			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
-		}
-		workflowRunTimeout := executionInfo.WorkflowRunTimeout
-		taskScheduleToStartTimeout = workflowRunTimeout
-	}
+	normalTaskQueue := mutableState.GetExecutionInfo().TaskQueue
```

Comment:
nit: you might use `normalTaskQueueName` here since it's not actually an instance of `TaskQueue`